### PR TITLE
Introduce keras-io documentation for KerasNLP

### PR DESCRIPTION
**Pull request by mattdangerw**
This still needs a few todos on the keras-nlp side:
 * [x] Land https://github.com/keras-team/keras-nlp/pull/132
 * [x] Make an 0.2.0 pip release.

As soon as the PR above lands I can cut a pre-release pip version and install it here so everything is runnable.

Feedback on the transformer guide appreciated!

### CODE REVIEW

File: templates/keras_nlp/index.md [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+---
+## Quick links
+
+* [Developer guides](/guides/keras_nlp/)
+* [API reference](/api/keras_nlp/)
+
+---
+## Installation
+
+KerasNLP requires **Python 3.7+** and **TensorFlow 2.9+**.
```
Comment (chenmoneygithub) [inline templates/keras_nlp/index.md:12]: Do we need TF 2.9+ for KerasNLP? I know CV is waiting for the `BaseAugmentationLayer`.

**Review by chenmoneygithub — COMMENTED** [review]
Have not reviewed the guide( transformer_pretraining.py), but other parts LGTM!

File: templates/keras_nlp/index.md (hunk 1 above)
Comment (fchollet) [reply templates/keras_nlp/index.md:12]: It has only been tested with 2.9 so this is a reasonable requirement.

File: guides/keras_nlp/transformer_pretraining.py [hunk 2]
```diff
@@ -0,0 +1,481 @@
...
+    .batch(PRETRAINING_BATCH_SIZE)
+)
+wiki_val_ds = (
+    tf.data.TextLineDataset("wikitext-103-raw/wiki.valid.raw")
+    .filter(lambda x: tf.strings.length(x) > 100)
+    .batch(PRETRAINING_BATCH_SIZE)
+)
+
+# Take a peak at the sst-2 dataset.
+list(sst_train_ds.unbatch().take(4).as_numpy_iterator())
```
Comment (fchollet) [inline guides/keras_nlp/transformer_pretraining.py:129]: Anything you want to print should be printed explicitly; this file should work as a Python script. Consider iterating over the dataset and calling `print()` on each sample

File: guides/keras_nlp/transformer_pretraining.py [hunk 3]
```diff
@@ -0,0 +1,481 @@
...
+        "tokens": outputs["tokens"],
+        "mask_positions": outputs["mask_positions"],
+    }
+    labels = outputs["mask_ids"]
+    weights = outputs["mask_weights"]
+    return features, labels, weights
+
+
+# We use prefetch() to pre-compute preprocessed batches on the fly on our CPU.
+pretrain_ds = wiki_train_ds.map(
```
Comment (fchollet) [inline guides/keras_nlp/transformer_pretraining.py:262]: How large is it? In some cases we may want to recommend `cache()` so that we only do the preprocessing work once.

File: guides/keras_nlp/transformer_pretraining.py [hunk 4]
```diff
@@ -0,0 +1,481 @@
...
+    preprocess, num_parallel_calls=tf.data.AUTOTUNE
+).prefetch(tf.data.AUTOTUNE)
+pretrain_val_ds = wiki_val_ds.map(
+    preprocess, num_parallel_calls=tf.data.AUTOTUNE
+).prefetch(tf.data.AUTOTUNE)
+
+# Preview a single input example.
+# The masks will change each time you run the cell.
+pretrain_val_ds.take(1).get_single_element()
```
Comment (fchollet) [inline guides/keras_nlp/transformer_pretraining.py:271]: Likewise, this should be printed explicitly

File: guides/keras_nlp/transformer_pretraining.py [hunk 5]
```diff
@@ -0,0 +1,481 @@
...
+# Predict an output word for each masked input token.
+# We use the input token embedding to project from our encoded vectors to
+# vocabulary logits, which has been shown to improve training efficiency.
+outputs = keras_nlp.layers.MLMClassificationHead(
+    embedding_weights=embedding_layer.token_embedding.embeddings,
+    outer_activation="softmax",
+)(encoded_tokens, mask_positions=inputs["mask_positions"])
+
+# Define and compile our pretraining model.
+pretraining_model = tf.keras.Model(inputs, outputs)
```
Comment (fchollet) [inline guides/keras_nlp/transformer_pretraining.py:367]: `keras` is imported, so just `keras.Model`

File: guides/keras_nlp/transformer_pretraining.py [hunk 6]
```diff
@@ -0,0 +1,481 @@
...
+
+# Take as input the tokenized input.
+inputs = keras.Input(shape=(SEQ_LENGTH,), dtype=tf.int32)
+
+# Encode and pool the tokens.
+encoded_tokens = encoder_model(inputs)
+pooled_tokens = keras.layers.GlobalAveragePooling1D()(encoded_tokens)
+
+# Predict an output label.
+outputs = tf.keras.layers.Dense(1, activation="sigmoid")(pooled_tokens)
```
Comment (fchollet) [inline guides/keras_nlp/transformer_pretraining.py:433]: `keras` is imported, so just `keras.layers.Dense`

File: guides/keras_nlp/transformer_pretraining.py [hunk 7]
```diff
@@ -0,0 +1,481 @@
...
+
+# Define and compile our finetuning model.
+finetuning_model = tf.keras.Model(inputs, outputs)
```
Comment (fchollet) [inline guides/keras_nlp/transformer_pretraining.py:436]: `keras` is imported, so just `keras.Model`

File: guides/keras_nlp/transformer_pretraining.py [hunk 8]
```diff
@@ -0,0 +1,481 @@
...
+layer. One of the key advantages of KerasNLP is all preprocessing is done inside the
+[tensorflow graph](https://www.tensorflow.org/guide/intro_to_graphs), making it possible
+to save and restore a model that can directly run inference on raw text!
+"""
+
+# Add our tokenization into our final model.
+inputs = keras.Input(shape=(), dtype=tf.string)
+tokens = tokenizer(inputs)
+outputs = finetuning_model(tokens)
+final_model = tf.keras.Model(inputs, outputs)
```
Comment (fchollet) [inline guides/keras_nlp/transformer_pretraining.py:469]: `keras` is imported, so just `keras.Model`

File: guides/keras_nlp/transformer_pretraining.py [hunk 9]
```diff
@@ -0,0 +1,481 @@
...
+final_model.save("final_model")
+
+# This model can predict directly on raw text.
+restored_model = tf.keras.models.load_model("final_model")
```
Comment (fchollet) [inline guides/keras_nlp/transformer_pretraining.py:473]: `keras` is imported, so just `keras.models.load_model`

File: scripts/nlp_api_master.py [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+        },
+        {
+            'path': 'unicode_character_tokenizer',
+            'title': 'UnicodeCharacterTokenizer',
+            'generate': ['keras_nlp.tokenizers.UnicodeCharacterTokenizer']
+        },
+    ]
+}
+
+PREPROCESSING_MASTER = {
```
Comment (fchollet) [inline scripts/nlp_api_master.py:29]: Should this exist? Why not just `layers` as a flat namespace?

**Review by fchollet — COMMENTED** [review]
Thanks for the PR. Haven't reviewed the text copy yet, just focused on the code.

File: guides/keras_nlp/transformer_pretraining.py (hunk 3 above)
Comment (mattdangerw) [reply guides/keras_nlp/transformer_pretraining.py:262]: Maybe 500mb?

But I don't think we can cache, because we want the masking to be a random augmentation not static. We could cache after tokenization, but I wouldn't want the complexity of breaking up the preprocess function. I can check if that's a speedup.

File: scripts/nlp_api_master.py (hunk 10 above)
Comment (mattdangerw) [reply scripts/nlp_api_master.py:29]: I think flat would probably be best too. Right now keras_nlp and keras_cv have preprocessing layers aliased and documented in two places. `keras_nlp.layers.preprocessing.X` and `keras_nlp.layers.X`. Ditto for cv.

Should probably touch on this one as a group.

File: guides/keras_nlp/transformer_pretraining.py (hunk 3 above)
Comment (fchollet) [reply guides/keras_nlp/transformer_pretraining.py:262]: Got it. Makes sense not to cache then.

File: scripts/nlp_api_master.py (hunk 10 above)
Comment (fchollet) [reply scripts/nlp_api_master.py:29]: Right, we went with a flat namespace for `keras.layers` as well. Deeply nested namespaces are user-hostile (as demonstrated by packages that practice them, like scipy) and can lead to classification dilemmas...

File: guides/keras_nlp/transformer_pretraining.py [hunk 11]
```diff
@@ -0,0 +1,481 @@
...
+- [WikiText-103](https://paperswithcode.com/dataset/wikitext-103): A medium sized
+collection of featured articles from English wikipedia, which we will use for
+pre-training.
+
+Finally, we will download a WordPiece vocabulary, to do sub-word tokenization later on in
+this guide.
+"""
+
+"""shell
+!pip install -q keras-nlp
```
[the remark this answers is not in the document]
Comment (LukeWood) [reply guides/keras_nlp/transformer_pretraining.py:42]: Does our """shell ...""" directive make these snippets work in Colab?

File: guides/keras_nlp/transformer_pretraining.py [hunk 12]
```diff
@@ -0,0 +1,481 @@
...
+
+"""
+### Establish a baseline
+
+As a first step, we will establish a baseline of good performance. We don't actually need
+KerasNLP for this, we can just use core Keras layers.
+
+We will build a simple bag-of-words model, where we learn a positive or negative weight
+for each word in our input. A sample's score is simply the sum of all words that are
+present in the sample.
```
Comment (chenmoneygithub) [inline guides/keras_nlp/transformer_pretraining.py:145]: sum of words' weights?

File: guides/keras_nlp/transformer_pretraining.py [hunk 13]
```diff
@@ -0,0 +1,481 @@
...
+"""
+
+# This layer will turn our input sentence into a list of 1s and 0s the same size
+# our vocabulary, indicating whether a word is present in absent.
+multi_hot_layer = keras.layers.TextVectorization(
+    max_tokens=4000, output_mode="multi_hot"
+)
+multi_hot_layer.adapt(sst_train_ds.map(lambda x, y: x))
+# We then  learn a linear regression over that layer, and that's our entire
```
Comment (chenmoneygithub) [inline guides/keras_nlp/transformer_pretraining.py:154]: nit: extra space

File: guides/keras_nlp/transformer_pretraining.py [hunk 14]
```diff
@@ -0,0 +1,481 @@
...
+baseline_model.fit(sst_train_ds, validation_data=sst_val_ds, epochs=5)
+
+"""
+A bag-of-words approach can be a fast and suprisingly powerful, especially when input
+examples contain a large number of words. With shorter sequences, it can hit a
+performance ceiling.
+
+To do better, we would like to build a model that can evaluate words *in context*. It's
+more useful to see the phrase "not my cup of tea" in an input, than to be told a sample
+contained the word "cup" or "tea" individually. Instead of evaluating each word in a
```
Comment (chenmoneygithub) [inline guides/keras_nlp/transformer_pretraining.py:171]: containing

File: guides/keras_nlp/transformer_pretraining.py [hunk 15]
```diff
@@ -0,0 +1,481 @@
...
+void, we need to use the information contained in the *entire ordered sequence* of our
+input.
+
+This runs us into a problem. `SST-2` is very small dataset, and there's simply not enough
+example text to attempt to build a larger, more parameterized, models that can learn on a
```
Comment (chenmoneygithub) [inline guides/keras_nlp/transformer_pretraining.py:176]: nit: remove ',' before "models that...". 

Also "models" should be in the singular form.

File: guides/keras_nlp/transformer_pretraining.py [hunk 16]
```diff
@@ -0,0 +1,481 @@
...
+
+To beat our baseline, we need to leverage the `WikiText103` dataset, an unlabeled
+collection of wikipedia articles that is much bigger than than `SST-2`.
+
+We are going to train a *transformer* model, a highly expressive model which will learn
+to embed each word in our input as a low dimentional vector. Our wikipedia dataset has no
+labels, so we will use an unsupervised training objective called the *Masked Language
+Modeling* (MLM) ojective.
+
+Essentially, we will be playing a big game of "guess the missing word." For each input
```
Comment (chenmoneygithub) [inline guides/keras_nlp/transformer_pretraining.py:196]: nit: move period out of "".

I am not actually sure about this though...

**Review by chenmoneygithub — CHANGES_REQUESTED** [review]
Thanks for the tutorial! It looks really nice!

Dropped some comments until the section "Preprocess data for the MLM task".

File: guides/keras_nlp/transformer_pretraining.py [hunk 17]
```diff
@@ -0,0 +1,481 @@
...
+
+We use `keras_nlp.layers.TokenAndPositionEmbedding` to first embed our input token ids.
+This layer simultaneously learns two embeddings--one for words in a sentence and another
+for integer positions in a sentence. The output embedding is simply the sum of the two.
+
+Then we can add a series of `keras_nlp.layers.TransformerEncoder` layers. These are the
+bread and butter of the transformer model, using an attention mechanism to attend to
+different parts of the input sentence, followed by a multi-layer perceptron block.
+
+The output of this model will be will be a encoded vector per input token id. Unlike the
```
Comment (chenmoneygithub) [inline guides/keras_nlp/transformer_pretraining.py:301]: duplicated "will be"

File: guides/keras_nlp/transformer_pretraining.py [hunk 18]
```diff
@@ -0,0 +1,481 @@
...
+bag-of-words model we used as a baseline, this model will embed each token accounting for
+the context in which in appeared.
```
Comment (chenmoneygithub) [inline guides/keras_nlp/transformer_pretraining.py:303]: => in which it appeared

File: guides/keras_nlp/transformer_pretraining.py [hunk 19]
```diff
@@ -0,0 +1,481 @@
...
+# We use prefetch() to pre-compute preprocessed batches on the fly on our CPU.
+finetune_ds = sst_train_ds.map(
+    preprocess, num_parallel_calls=tf.data.AUTOTUNE
+).prefetch(tf.data.AUTOTUNE)
+finetune_val_ds = sst_val_ds.map(
+    preprocess, num_parallel_calls=tf.data.AUTOTUNE
+).prefetch(tf.data.AUTOTUNE)
+
+# Preview a single input example.
+finetune_val_ds.take(1).get_single_element()
```
Comment (chenmoneygithub) [inline guides/keras_nlp/transformer_pretraining.py:412]: Explicit print here.

**Review by chenmoneygithub — COMMENTED** [review]
comments on the second half.

File: guides/keras_nlp/transformer_pretraining.py (hunk 16 above)
Comment (mattdangerw) [reply guides/keras_nlp/transformer_pretraining.py:196]: Yeah, probably debatable, but it's a consistent style for this guide at least. Will do.

For technical writing, I think the "quote." style is terrible.

File: guides/keras_nlp/transformer_pretraining.py [hunk 20]
```diff
@@ -0,0 +1,481 @@
...
+every word in your corpus, but you can still encode common words in their entirety.
+
+The second thing we need to do is mask our input for the MLM task. To do this, we can use
+`keras_nlp.layers.MLMMaskGenerator`, which will randomly select a set of tokens in each
+input and mask them out.
+
+The tokenizer and the masking layer can both be used inside a call to
+[tf.data.Dataset.map](https://www.tensorflow.org/api_docs/python/tf/data/Dataset#map).
+This will allow us to efficiently pre-compute each batch on the CPU, while our GPU or TPU
+works on training with the batch that came before. And because our masking layer will
```
[the remark this answers is not in the document]
Comment (mattdangerw) [reply guides/keras_nlp/transformer_pretraining.py:227]: I think "simplified BERT" is more likely to mean something to a random person than "RoBERTa style".

File: guides/keras_nlp/transformer_pretraining.py [hunk 21]
```diff
@@ -0,0 +1,481 @@
...
+"""
+
+# Setting sequence_length will trim or pad the token outputs to shape
+# (batch_size, SEQ_LENGTH).
+tokenizer = keras_nlp.tokenizers.WordPieceTokenizer(
+    vocabulary="bert_vocab_uncased.txt", sequence_length=SEQ_LENGTH,
+)
+# Setting mask_selection_length will trim or pad the mask outputs to shape
+# (batch_size, PREDICTIONS_PER_SEQ).
+masker = keras_nlp.layers.MLMMaskGenerator(
```
[the remark this answers is not in the document]
Comment (mattdangerw) [reply guides/keras_nlp/transformer_pretraining.py:234]: We print the sorted tuple of output later (and it's a pretty bulky print). I don't I think we should print it twice, it would take up too much vertical space.

**Comment by mattdangerw** [comment]
Thanks for review! Address most of the comments and pinned the keras-nlp version to a pre-release so it's testable.

Still need to switch to python file downloads, and and figure out if we are getting rid of the `preprocessing` namespace.

**Review by LukeWood — APPROVED** [review]
I think it is looking good!

File: guides/keras_nlp/transformer_pretraining.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (chenmoneygithub) [reply guides/keras_nlp/transformer_pretraining.py:223]: makes sense!

File: scripts/nlp_api_master.py (hunk 10 above)
Comment (mattdangerw) [reply scripts/nlp_api_master.py]: Removed layers.preprocessing here.

**Review by fchollet — APPROVED** [review]
LGTM. Let's merge and if there are any remaining issues let's fix them forward.
